fms_dgt/base/block.py

```diff
@@ -0,0 +1,136 @@
+        name: str = None,
+        arg_fields: List[str] = None,
+        kwarg_fields: List[str] = None,
+        result_field: str = None,
+    ) -> None:
+
+        self._name = name
+
+        # minor type checking
+        if type(arg_fields) == str:
+            arg_fields = [arg_fields]
+        if type(kwarg_fields) == str:
+            kwarg_fields = [kwarg_fields]
+        if type(result_field) == list:
+            assert (
+                len(result_field) == 1
+            ), "Cannot have multiple 'result' fields for {name}"
+            result_field = result_field[0]
+
+        self._arg_fields = arg_fields
+        self._kwarg_fields = kwarg_fields
+        self._result_field = result_field
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def arg_fields(self):
+        return self._arg_fields
+
+    @property
+    def kwarg_fields(self):
+        return self._kwarg_fields
+
+    @property
+    def result_field(self):
+        return self._result_field
+
+    def generate(
+        self,
+        inputs: BLOCK_INPUT_TYPE,
+        arg_fields: Optional[List[str]] = None,
+        kwarg_fields: Optional[List[str]] = None,
+        result_field: Optional[str] = None,
+    ):
+        raise NotImplementedError
+
+
+class BaseUtilityBlock(BaseBlock):
+    pass
+
+
+class BaseGeneratorBlock(BaseBlock):
+    pass
+
+
+class BaseValidatorBlock(BaseBlock):
+    def __init__(self, filter: bool = False, **kwargs: Any) -> None:
+        super().__init__(**kwargs)
+        self._filter_invalids = filter
+
+    def generate(
+        self,
+        inputs: BLOCK_INPUT_TYPE,
+        arg_fields: Optional[List[str]] = None,
+        kwarg_fields: Optional[List[str]] = None,
+        result_field: Optional[List[str]] = None,
+    ):
+        outputs = []
+        for x in inputs:
+            inp_args, inp_kwargs = get_args_kwargs(
+                x, arg_fields or self.arg_fields, kwarg_fields or self.kwarg_fields
+            )
+            res = self._validate(*inp_args, **inp_kwargs)
+            if res or not self._filter_invalids:
+                write_result(x, res, result_field or self.result_field)
+                outputs.append(x)
+        return outputs
+
+    def _validate(self, *args: Any, **kwargs: Any) -> bool:
+        raise NotImplementedError
+
+
+def get_args_kwargs(
+    inp: BLOCK_ROW_TYPE,
+    arg_fields: Optional[List[str]] = None,
+    kwarg_fields: Optional[List[str]] = None,
+):
+    if arg_fields is None:
+        arg_fields = []
+    if kwarg_fields is None:
+        kwarg_fields = []
+
+    if type(inp) == dict:
+        args = [inp.get(arg) for arg in arg_fields]
+        kwargs = {kwarg: inp.get(kwarg) for kwarg in kwarg_fields}
+    elif type(inp) in [pd.DataFrame, Dataset]:
+        args = [inp.get(arg) for arg in arg_fields]
+        kwargs = {kwarg: inp.get(kwarg) for kwarg in kwarg_fields}
+    else:
+        raise ValueError(f"Unexpected input type: {type(inp)}")
```

Comment:
Two things that I think would be cleaner here:

1. It looks like the body of the `if` and `elif` clauses are the same, so they can be collapsed
2. This should use the more idiomatic `isinstance(x, types)` type checking so that subclasses are respected

```suggestion
    if isinstance(inp, (dict, pd.DataFrame, Dataset):
        args = [inp.get(arg) for arg in arg_fields]
        kwargs = {kwarg: inp.get(kwarg) for kwarg in kwarg_fields}
    else:
        raise ValueError(f"Unexpected input type: {type(inp)}")
```

Reply:
Ah, yes good catch, I had initially put that second elif statement in there as a placeholder. But now revisiting it, I suspect we'll just handle it in the same way as the if statement, so collapsing the two is good